src/wasm/wasm-binary.cpp

```diff
@@ -1597,20 +1597,23 @@ void WasmBinaryBuilder::readNames(size_t payloadLen) {
       if (import->kind != ExternalKind::Function) continue;
       importedFunctions++;
     }
-    std::set<Name> functionNames;
     for (size_t i = 0; i < num; i++) {
       auto index = getU32LEB();
       if (index < importedFunctions) {
         getInlineString(); // TODO: use this
       } else if (index - importedFunctions < functions.size()) {
         auto name = getInlineString();
         functions[index - importedFunctions]->name = name;
-        if (functionNames.count(name) > 0) {
-          throw ParseException("duplicate function names");
-        }
-        functionNames.insert(name);
       }
     }
+    // disallow duplicate names
+    std::set<Name> functionNames;
+    for (auto* func : functions) {
+      if (functionNames.count(func->name) > 0) {
+        throw ParseException("duplicate function name: " + std::string(func->name.str));
+      }
+      functionNames.insert(func->name);
```

Comment:
How about, instead of calling `count` followed by `insert`,
 just calling `insert`, and then throwing if `result.second` is false? That way we only do one lookup instead of 2.